src/test/util/setup_common.cpp

```diff
@@ -120,18 +120,53 @@ BasicTestingSetup::BasicTestingSetup(const ChainType chainType, const std::vecto
         arguments = Cat(arguments, G_TEST_COMMAND_LINE_ARGUMENTS());
     }
     util::ThreadRename("test");
-    fs::create_directories(m_path_root);
-    m_args.ForceSetArg("-datadir", fs::PathToString(m_path_root));
-    gArgs.ForceSetArg("-datadir", fs::PathToString(m_path_root));
     gArgs.ClearPathCache();
     {
         SetupServerArgs(*m_node.args);
+        m_node.args->AddArg("-testdatadir", strprintf("Custom data directory (default: %s<random_string>)", fs::PathToString(fs::temp_directory_path() / "test_common_" PACKAGE_NAME / "")),
+                            ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
         std::string error;
         if (!m_node.args->ParseParameters(arguments.size(), arguments.data(), error)) {
             m_node.args->ClearArgs();
             throw std::runtime_error{error};
         }
     }
+
+    if (!m_node.args->IsArgSet("-testdatadir")) {
+        // By default, the data directory has a random name
+        const auto rand_str{g_insecure_rand_ctx_temp_path.rand256().ToString()};
+        m_path_root = fs::temp_directory_path() / "test_common_" PACKAGE_NAME / rand_str;
+        TryCreateDirectories(m_path_root);
+    } else {
+        // Custom data directory
+        m_has_custom_datadir = true;
+        fs::path root_dir{m_node.args->GetPathArg("-testdatadir")};
+        if (root_dir.empty()) {
+            std::cerr << "-testdatadir argument is empty, please specify a path\n";
+            exit(EXIT_FAILURE);
+        }
+        const std::string test_path{G_TEST_GET_FULL_NAME ? G_TEST_GET_FULL_NAME() : ""};
+        const fs::path lockdir{root_dir / "test_temp" / fs::PathFromString(test_path)};
+        m_path_root = lockdir / "datadir";
+
+        // Try to obtain the lock; if unsuccessful don't disturb the existing test.
+        TryCreateDirectories(lockdir);
+        if (util::LockDirectory(lockdir, ".lock", /*probe_only=*/false) != util::LockResult::Success) {
+            std::cerr << "Cannot obtain a lock on test data lock directory " + fs::PathToString(lockdir) + '\n' +
+                             "The test executable is probably already running.\n";
+            exit(EXIT_FAILURE);
+        }
+
+        // Always start with a fresh data directory; this doesn't delete the .lock file.
+        fs::remove_all(m_path_root);
+        TryCreateDirectories(m_path_root);
+
+        // Print the test directory name if custom.
+        std::cout << "Test directory (will not be deleted): " << m_path_root << std::endl;
```

Comment:
Suggestion: change to something like `"Test directory (not deleted until next run): "`

Reply:
Thanks, I _think_ it's clear enough as is. If it says "until next run", it might not be clear if that's before the next run of _any_ test or this _particular_ test. Clarifying that would add more text and be a bit too chatty. I think the current text is clear enough to convey "not deleted after the current run"; I think it's understood that if you specify the same data directory and the same test, it will overwrite that same directory.

Reply:
Last suggestion - inspired by your response: `"Test directory (remains after run)"`.

If it hadn't been so chatty I compared to running without `-testdatadir` I wouldn't mind. Would have been nicer to explain the part within the parenthesis one single time, before all the tests are run.

Reply:
> Last suggestion: `"Test directory (remains after run)"`.

I'm not sure if I see that as an improvement; I'll hold off to see if anyone else has a preference.

> Would have been nicer to explain the part within the parenthesis one single time, before all the tests are run.

A previous iteration of this PR did this (before each test case had its own data directory), the code worked but was messy (required a static (global) variable), so I was glad to remove it which needed to be done anyway since now each test has its own data directory. I don't really want to add that complexity back in. I think I'll leave it as is unless someone has a strong opinion to the contrary.